internal/pkg/agent/vault/seed.go

```diff
@@ -16,62 +17,116 @@ import (
 )
 
 const (
+	// seedFile is len(aesgcm.AES256) and contains only the random seed
+	// A default salt size of 8 is used with this seed file
 	seedFile = ".seed"
+	// seedFileV2 is len(aesgcm.AES256)+4 and contains the random seed followed by a non-zero salt size (little endian uint32)
+	seedFileV2     = ".seedV2"
+	seedFileV2Size = int(aesgcm.AES256) + 4
+)
+
+const (
+	saltSizeV1        = 8
+	defaultSaltSizeV2 = 16
 )
 
 var (
 	mxSeed sync.Mutex
 )
 
-func getSeed(path string) ([]byte, error) {
-	fp := filepath.Join(path, seedFile)
-
+func getSeed(path string) ([]byte, int, error) {
 	mxSeed.Lock()
 	defer mxSeed.Unlock()
 
+	// Prefer V2 seeds
+	b, saltSize, errV2 := getSeedV2(path)
+	if errV2 == nil {
+		return b, saltSize, nil
+	}
+	// Fallback to V1 seed
+	b, errV1 := getSeedV1(path)
+	if errV1 == nil {
+		return b, saltSizeV1, nil
+	}
+	return nil, 0, errors.Join(errV2, errV1)
+}
+
+// getSeedV2 will read a seedV2 file and return the passphrase and saltSize
+// Will return fs.ErrNotExists if the byte count does not match, or saltSize is 0
+// when in FIPS mode will return fs.ErrUnsupported when saltSize is non-zero but less then 16
+func getSeedV2(path string) ([]byte, int, error) {
+	fp := filepath.Join(path, seedFileV2)
+
 	b, err := os.ReadFile(fp)
 	if err != nil {
-		return nil, fmt.Errorf("could not read seed file: %w", err)
+		return nil, 0, fmt.Errorf("could not read seed file: %w", err)
 	}
 
-	// return fs.ErrNotExists if invalid length of bytes returned
-	if len(b) != int(aesgcm.AES256) {
-		return nil, fmt.Errorf("invalid seed length, expected: %v, got: %v: %w", int(aesgcm.AES256), len(b), fs.ErrNotExist)
+	// return fs.ErrNotExist if invalid length of bytes returned
+	if len(b) != seedFileV2Size {
+		return nil, 0, fmt.Errorf("invalid seed length, expected: %v, got: %v: %w", seedFileV2Size, len(b), fs.ErrNotExist)
+	}
+	pass := b[0:int(aesgcm.AES256)]
+	saltSize := binary.LittleEndian.Uint32(b[int(aesgcm.AES256):])
+	if saltSize == 0 {
+		return nil, 0, fmt.Errorf("salt size 0 detected: %w", fs.ErrNotExist)
 	}
-	return b, nil
+	if err := checkSalt(int(saltSize)); err != nil {
+		return nil, 0, err
+	}
+	return pass, int(saltSize), nil
 }
 
-func createSeedIfNotExists(path string) ([]byte, error) {
-	fp := filepath.Join(path, seedFile)
-
+func createSeedIfNotExists(path string) ([]byte, int, error) {
 	mxSeed.Lock()
 	defer mxSeed.Unlock()
 
-	b, err := os.ReadFile(fp)
+	// Prefer reading V2 seeds
+	pass, saltSize, err := getSeedV2(path)
 	if err != nil {
 		if !errors.Is(err, os.ErrNotExist) {
-			return nil, err
+			return nil, 0, err
 		}
 	}
+	if len(pass) != 0 {
+		return pass, saltSize, nil
+	}
 
-	if len(b) != 0 {
-		return b, nil
+	// V1 seed fallback
+	// getSeedV1 will return ErrNotExist when in FIPS mode.
+	pass, err = getSeedV1(path)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return nil, 0, err
+		}
+	}
+	if len(pass) != 0 {
+		return pass, saltSizeV1, nil
```

Comment:
Based on my reading here this means that if v2 seed is not present and only v1 is present then it will only read the v1 seed and never create the v2 seed. I am reading that correctly, yes?

How does that work when upgrading from a non-fips to a fips agent? Does it re-create the vault with a new large seed and move all of the data over from non-fips to fips compliant?

Reply:
>I will pose the question, should we support upgrades from a non-fips to a fips agent? 

+1 on this question; I believe @nimarezainia already confirmed in a different conversation that this is not required - @nimarezainia keep me honest here.

Reply:
I'll remove all v2 support from non-fips agents

Reply:
yes agreed. I can't see a scenario where the user goes between the two flavours. That is not to say someone won;t come up with one (complaining they don;t have access to the machines to do a reinstall). I think instinctively users would understand that fip is a new binary and needs a new install.  For Fleet managed agents, upgrading to a different falvor is not even possible. For standalone users the 'upgrade' command doesn't allow a different binary to be applied. 

I would say it's safe to assume moving to a fips enabled image needs a new installation.